go/cmd/dolt/commands/log.go

```diff
@@ -126,44 +146,153 @@ func (cmd LogCmd) logWithLoggerFunc(ctx context.Context, commandStr string, args
-			if err != nil {
-				cli.PrintErrln(color.HiRedString("invalid commit %s\n", apr.Arg(0)))
+	return opts, nil
+}
+
+func parseRefsAndTable(ctx context.Context, apr *argparser.ArgParseResults, dEnv *env.DoltEnv) (*doltdb.CommitSpec, *doltdb.CommitSpec, string, error) {
+	switch apr.NArg() {
+	// dolt log
+	case 0:
+		return nil, nil, "", nil
+
+	// dolt log <ref/^ref/revision-range/table>
+	case 1:
+		return getCommitOrTableFromString(ctx, apr.Arg(0), dEnv, true)
+
+	// dolt log <ref/^ref> <ref/^ref/table>
+	case 2:
+		firstCs, firstExNotCs, _, err := getCommitOrTableFromString(ctx, apr.Arg(0), dEnv, false)
+		if err != nil {
+			return nil, nil, "", err
+		}
+
+		secondCs, secondExNotCs, tableName, err := getCommitOrTableFromString(ctx, apr.Arg(1), dEnv, false)
+		if err != nil {
+			return nil, nil, "", err
+		}
+
+		if len(tableName) > 0 {
+			if firstExNotCs != nil {
+				return nil, nil, "", fmt.Errorf("Providing tableName for two dot log not yet supported")
 			}
-			return logCommits(ctx, dEnv, cs, opts)
-		} else {
-			return handleErrAndExit(logTableCommits(ctx, dEnv, opts, dEnv.RepoStateReader().CWBHeadSpec(), apr.Arg(0)))
+			// dolt log <ref> table
+			return firstCs, nil, tableName, nil
+		}
+
+		if firstCs != nil && secondCs != nil {
+			return nil, nil, "", fmt.Errorf("Cannot provide two commit refs") // dolt log <ref> <ref>
+		}
+		if firstExNotCs != nil && secondExNotCs != nil {
+			return nil, nil, "", fmt.Errorf("Cannot exclude two commit refs") // dolt log ^<ref> ^<ref>
+		}
+		// dolt log ^<ref> <ref>
+		if firstExNotCs != nil && secondCs != nil {
+			return secondCs, firstExNotCs, "", nil
+		}
+		// dolt log <ref> ^<ref>
+		if firstCs != nil && secondExNotCs != nil {
+			return firstCs, secondExNotCs, "", nil
+		}
+
+		return nil, nil, "", nil
+
+	default:
+		return nil, nil, "", fmt.Errorf("Cannot provide more than 3 arguments")
+	}
+}
+
+func getCommitOrTableFromString(ctx context.Context, str string, dEnv *env.DoltEnv, canDot bool) (*doltdb.CommitSpec, *doltdb.CommitSpec, string, error) {
+	// <ref>...<ref>
+	if strings.Contains(str, "...") {
+		return nil, nil, "", fmt.Errorf("Three dot dolt log not supported")
+	}
+
+	// <ref>..<ref>
+	if strings.Contains(str, "..") {
+		if !canDot {
+			return nil, nil, "", fmt.Errorf("Cannot use two dot when 2 arguments provided")
+		}
+		refs := strings.Split(str, "..")
+		notCs, err := getCommitSpec(refs[0])
+		if err != nil {
+			return nil, nil, "", err
+		}
+
+		cs, err := getCommitSpec(refs[1])
+		if err != nil {
+			return nil, nil, "", err
+		}
+
+		return cs, notCs, "", nil
+	}
+
+	// ^<ref>
+	if strings.Contains(str, "^") {
+		commit := strings.TrimPrefix(str, "^")
+		notCs, err := getCommitSpec(commit)
+		if err != nil {
+			return nil, nil, "", err
 		}
-	} else { // dolt log ref table
-		cs, err := doltdb.NewCommitSpec(apr.Arg(0))
+		return nil, notCs, "", err
+	}
+
+	argIsRef := actions.ValidateIsRef(ctx, str, dEnv.DoltDB, dEnv.RepoStateReader())
```

Comment:
There's probably some interesting edge cases here where we can't disambiguate between a branch named "branch1" used as a ref and a table named "branch1" used as a table. Probably worth thinking about and trying a test or two and even adding a test and documenting the behavior that when there is ambiguity between a ref/branch name and a table name that we'd assume a branch ref.

Reply:
Added some bats for this [here](https://github.com/dolthub/dolt/pull/4561/files#diff-90a77217a84d199ff6b5f127e1b52cf4dde4df409835e09e925c51a8023a2a25R125). If there's a branch and table named `myname`, dolt log will have these defaults:
```
dolt log myname[branch]
dolt log myname[branch] myname[table]
```
That means `dolt log myname[table]` won't work unless you provide a revision